src/table/use-sticky-columns.ts

```diff
@@ -0,0 +1,362 @@
+  stickyColumnsFirst: number;
+  stickyColumnsLast: number;
+}
+
+export default class StickyColumnsStore extends AsyncStore<StickyColumnsState> {
+  private cellOffsets = new Map<ColumnId, { first: number; last: number }>();
+  private stickyWidthLeft = 0;
+  private stickyWidthRight = 0;
+  private isStuckToTheLeft = false;
+  private isStuckToTheRight = false;
+
+  constructor() {
+    super({ cellState: {}, scrollPadding: { left: 0, right: 0 } });
+  }
+
+  public updateCellStyles(props: UpdateCellStylesProps) {
+    const hasStickyColumns = props.stickyColumnsFirst + props.stickyColumnsLast > 0;
+    const hadStickyColumns = this.cellOffsets.size > 0;
+
+    if (hasStickyColumns || hadStickyColumns) {
+      this.updateScroll(props);
+      this.updateCellOffsets(props);
+      this.set(() => ({
+        cellState: this.generateCellStyles(props),
+        scrollPadding: { left: this.stickyWidthLeft, right: this.stickyWidthRight },
+      }));
+    }
+  }
+
+  private updateScroll(props: UpdateCellStylesProps) {
+    const wrapperScrollLeft = props.wrapper.scrollLeft;
+    const wrapperScrollWidth = props.wrapper.scrollWidth;
+    const wrapperClientWidth = props.wrapper.clientWidth;
+    const tablePaddingLeft = parseInt(getComputedStyle(props.table).paddingLeft) || 0;
+    const tablePaddingRight = parseInt(getComputedStyle(props.table).paddingRight) || 0;
+
+    this.isStuckToTheLeft = wrapperScrollLeft > tablePaddingLeft;
+    this.isStuckToTheRight = Math.ceil(wrapperScrollLeft) < wrapperScrollWidth - wrapperClientWidth - tablePaddingRight;
+  }
+
+  private generateCellStyles = (props: UpdateCellStylesProps): Record<ColumnId, null | StickyColumnsCellState> => {
+    const isEnabled = this.isEnabled(props);
+
+    return props.visibleColumns.reduce((acc, columnId, index) => {
+      let stickySide = 'non-sticky';
+      if (index < props.stickyColumnsFirst) {
+        stickySide = 'left';
+      } else if (index >= props.visibleColumns.length - props.stickyColumnsLast) {
+        stickySide = 'right';
+      }
+
+      if (!isEnabled || stickySide === 'non-sticky') {
+        acc[columnId] = null;
+        return acc;
+      }
+
+      // Determine the offset of the sticky column using the `cellOffsets` state object
+      const isFirstColumn = index === 0;
+      const stickyColumnOffsetLeft = this.cellOffsets.get(columnId)?.first ?? 0;
+      const stickyColumnOffsetRight = this.cellOffsets.get(columnId)?.last ?? 0;
+      const cellStyle = {
+        left: stickyColumnOffsetLeft !== undefined && stickySide === 'left' ? stickyColumnOffsetLeft : undefined,
+        right: stickyColumnOffsetRight !== undefined && stickySide === 'right' ? stickyColumnOffsetRight : undefined,
+      };
+      const lastLeftStickyColumnIndex = props.stickyColumnsFirst - 1;
+      const lastRightStickyColumnIndex = props.visibleColumns.length - props.stickyColumnsLast;
+      const tablePaddingLeft = parseInt(getComputedStyle(props.table).paddingLeft) || 0;
```

Comment:
`getComputedStyle` can force the browser to relayout the page i. I'm not sure if the result of that is cached, but since this value should be the same for all cells I suggest to move this constant out of the loop, e.g. into line 274, to reduce the risk of a performance impact.

See also https://gist.github.com/paulirish/5d52fb081b3570c81e3a for more background

Reply:
Good point 👍

Reply:
Good call, also great linked resource. Thanks!